File: Graph Theory Peaks/Graph/Q1. Minimum Number of Vertices to Reach All Nodes/Q1_MinimumNumberofVerticestoReachAllNodes.py
class Solution(object):
    def findSmallestSetOfVertices(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        
        # range(n) → creates all nodes from 0 to n-1
        # set(range(n)) → convert all nodes into a set
        
        # (j for i, j in edges) → take only destination nodes from edges
        # set(j for i, j in edges) → set of all nodes that have incoming edges
        
        # Subtract destination nodes from all nodes
        # Remaining nodes are those with indegree = 0
        
        # Convert the result back to list and return
        return list(set(range(n)) - set(j for i, j in edges))

File: Graph Theory Peaks/Graph/Q1. Minimum Number of Vertices to Reach All Nodes/Q1_MinimumNumberofVerticestoReachAllNodes.java
class Solution {
    
    public List<Integer> findSmallestSetOfVertices(int n, List<List<Integer>> edges) {
        
        // Create an array to store indegree (number of incoming edges) for each node
        int[] indegree = new int[n];
        
        // Traverse each edge in the graph
        for (List<Integer> edge : edges) {
            
            // edge.get(0) = source node
            // edge.get(1) = destination node
            
            // Increase indegree of destination node
            indegree[edge.get(1)]++;
        }

        // List to store nodes with indegree 0
        List<Integer> ans = new ArrayList<>();
        
        // Check every node from 0 to n-1
        for (int i = 0; i < n; i++) {
            
            // If a node has no incoming edges,
            // it means no other node points to it
            if (indegree[i] == 0) {
                
                // So we must include it in the answer
                ans.add(i);
            }
        }

        // Return the list of nodes with indegree 0
        return ans;
    }
}s